### PR TITLE
Adding support for XLIO socketxtreme and callback-api

### PR DESCRIPTION
**Pull request by AlexanderGrissik**
1. Adding support for XLIO socketxtreme and callback-api.
2.1 Includes support for multithreading with socketxtreme
3. Moving generally unused options from template parameter to unlikely if condition. These options are not for performance anyway.
2.1 Reduced binary size from 200GB to 32GB
2.2 Reduced compilation time significantly.
2.3 No performance impact was observed
4. Adding warnings for incompatible options

### CODE REVIEW

File: src/sockperf.cpp [hunk 1]
```diff
@@ -3575,8 +3578,12 @@ int bringup(const int *p_daemonize) {
     }
 #endif // USING_EXTRA_API
 
-
 #if defined(DEFINED_TLS)
+    if (s_user_params.is_rxfiltercb || s_user_params.is_zcopyread || s_user_params.fd_handler_type == SOCKETXTREME) {
+        errno = EPERM;
+        exit_with_err("--tls is incompatible with VMA/XLIO Extra APIs", SOCKPERF_ERR_FATAL);
+    }
+
```
Comment (igor-ivanov) [inline src/sockperf.cpp:3589]: Please consider do this check in place as https://github.com/Mellanox/sockperf/blob/71490a052cbc20bf1799413ed784966473be23e7/src/sockperf.cpp#L2208

File: src/sockperf.cpp [hunk 2]
```diff
@@ -1527,8 +1527,8 @@ static int proc_mode_server(int id, int argc, const char **argv) {
         { OPT_RXFILTERCB,
           AOPT_NOARG,
           aopt_set_literal(0),
-          aopt_set_string("rxfiltercb", "vmarxfiltercb"),
-          "Use VMA's receive path message filter callback API (See VMA's readme)." },
+          aopt_set_string("rxfiltercb"),
+          "Use extra API receive path message filter callback API (See VMA/XLIO readme)." },
```
Comment (igor-ivanov) [inline src/sockperf.cpp:1530]: could you update `OPT_RATE_LIMIT` option with `VMA/XLIO` and `VMA_RING_ALLOCATION_LOGIC_TX/XLIO_RING_ALLOCATION_LOGIC_TX` changes

File: src/sockperf.cpp [hunk 3]
```diff
@@ -3511,16 +3547,9 @@ int bringup(const int *p_daemonize) {
...
-            if (s_user_params.is_rxfiltercb || s_user_params.fd_handler_type == SOCKETXTREME) {
-                exit_with_err("VMA Extra API is not available", SOCKPERF_ERR_FATAL);
-            }
-        } else {
+        if (g_vma_api) { // Try VMA Extra API
             log_msg("VMA Extra API is in use");
-        }
-
-        if (!g_vma_api) { // Try XLIO
+        } else { // Try XLIO
```
Comment (igor-ivanov) [inline src/sockperf.cpp:3547]: please consider change this code to be more symmetric for vma and xlio. I propose to do
```
#ifdef USING_VMA_EXTRA_API
        g_vma_api = vma_get_api();
#endif // USING_VMA_EXTRA_API
        if (g_vma_api) { // Try VMA Extra API
            log_msg("VMA Extra API is in use");
        } else {
#ifdef USING_XLIO_EXTRA_API
            g_xlio_api = xlio_get_api();
#endif // USING_XLIO_EXTRA_API
            if (!g_xlio_api) {
             log_msg("XLIO Extra API is in use");
       }
        if (!g_vma_api && !g_xlio_api) {
            errno = EPERM;
            exit_with_err("VMA or XLIO Extra API is not available", SOCKPERF_ERR_FATAL);
        }
```

File: src/sockperf.cpp (hunk 1 above)
Comment (AlexanderGrissik) [reply src/sockperf.cpp:3589]: The thing is that not all vars are under common_obj.

File: src/sockperf.cpp (hunk 3 above)
Comment (AlexanderGrissik) [reply src/sockperf.cpp:3547]: Done

File: src/sockperf.cpp (hunk 2 above)
Comment (AlexanderGrissik) [reply src/sockperf.cpp:1530]: Done

**Review by igor-ivanov — APPROVED** [review]

**Pull request merged**